core/src/main/scala/dagr/core/execsystem/TaskManagerLike.scala

```diff
@@ -0,0 +1,177 @@
+
+  /** Checks if we have failed tasks.
+    *
+    * @return true if we have failed tasks, false otherwise.
+    */
+  def hasFailedTasks: Boolean
+
+  /** Adds a task to be managed
+    *
+    * @param task the given task.
+    * @return the task identifier.
+    */
+  def addTask(task: Task): BigInt
+
+  /** Adds tasks to be managed
+    *
+    * @param tasks the given tasks.
+    * @return the task identifiers.
+    */
+  def addTasks(tasks: Task*): List[BigInt]
+
+  /** Resubmit a task for execution.  This will stop the task if it is currently running, and queue
+    * it up for execution.  The number of attempts will be reset to zero.
+    *
+    * @param task the task to resubmit.
+    * @return true if the task was successfully resubmitted, false otherwise.
+    */
+  // Turning it off until `resubmit` is used.
+  //def resubmitTask(task: Task): Boolean
+
+  /** Replace the original task with the replacement task and update any internal references.
+    *
+    * If the replacement task depends on any task, any tasks depends on the replacement task, or if the original
+    * task is not being tracked this method will return false.  Any tasks dependent on the original task, or any
+    * task on which the original task depends, will have their dependency relationship updated.
+    *
+    * @param original the original task to replace.
+    * @param replacement the replacement task.
+    * @return true if we are successful, false otherwise.
+    */
+  def replaceTask(original: Task, replacement: Task): Boolean
+
+  /** Run all tasks managed to either completion, failure, or inability to schedule.
+    *
+    * This will terminate tasks that were still running before returning in the case of failure or
+    * inability to schedule.
+    *
+    * @param sleepMilliseconds the time to wait in milliseconds to wait between trying to schedule tasks.
+    * @param timeout           the length of time in milliseconds to wait for running tasks to complete
+    * @return a bi-directional map from the set of tasks to their execution information.
+    */
+  def runToCompletion(sleepMilliseconds: Int = 1000, timeout: Int = 1000): BiMap[Task, TaskExecutionInfo]
+
+  /** Run a a single iteration of managing tasks.
+    *
+    * 1. Get tasks that have completed and update their state.
+    * 2. Update any resolved dependencies in the execution graph.
+    * 3. Get tasks for any task that has no predecessors until no more can be found.
+    * 4. Schedule tasks and start running them.
+    *
+    * @param timeout the length of time in milliseconds to wait for running tasks to complete
+    * @return a tuple of:
+    *         (1) tasks that can be scheduled.
+    *         (2) tasks that were scheduled.
+    *         (3) tasks that are running prior to scheduling.
+    *         (4) the tasks that have completed prior to scheduling.
+    */
+  def stepExecution(timeout: Int = 1000): (List[Task], List[Task], List[Task], List[Task])
```

Comment:
Just re-iterating that if these could be `Iterable` or `Traversable` that might be friendlier.